pkg/kn/commands/service/import.go

```diff
@@ -0,0 +1,143 @@
+func NewServiceImportCommand(p *commands.KnParams) *cobra.Command {
+	var waitFlags commands.WaitFlags
+
+	command := &cobra.Command{
+		Use:   "import FILENAME",
+		Short: "Import a service and its revisions",
+		Example: `
+ # Import a service from YAML file
+ kn service import /path/to/file.yaml
+
+ # Import a service from JSON file
+ kn service import /path/to/file.json`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return errors.New("'kn service import' requires filename of import file as single argument")
+			}
+			filename := args[0]
+
+			namespace, err := p.GetNamespace(cmd)
+			if err != nil {
+				return err
+			}
+
+			client, err := p.NewServingClient(namespace)
+			if err != nil {
+				return err
+			}
+
+			return importWithOwnerRef(client, filename, cmd.OutOrStdout(), waitFlags)
+		},
+	}
+	flags := command.Flags()
+	commands.AddNamespaceFlags(flags, false)
+	waitFlags.AddConditionWaitFlags(command, commands.WaitDefaultTimeout, "import", "service", "ready")
+
+	return command
+}
+
+func importWithOwnerRef(client clientservingv1.KnServingClient, filename string, out io.Writer, waitFlags commands.WaitFlags) error {
+	var export clientv1alpha1.Export
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 512)
+	err = decoder.Decode(&export)
+	if err != nil {
+		return err
+	}
+	if export.Spec.Service.Name == "" {
+		return fmt.Errorf("provided import file doesn't contain service name, please note that only kn' custom Export format is supported")
+	}
+
+	serviceName := export.Spec.Service.Name
+
+	// Return error if service already exists
+	svcExists, err := serviceExists(client, serviceName)
+	if err != nil {
+		return err
+	}
+	if svcExists {
+		return fmt.Errorf("cannot import service '%s' in namespace '%s' because the service already exists",
+			serviceName, client.Namespace())
+	}
+
+	err = client.CreateService(&export.Spec.Service)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve current Configuration to be use in OwnerReference
+	currentConf, err := getConfigurationWithRetry(client, serviceName)
+	if err != nil {
+		return err
+	}
+
+	// Create revision with current Configuration's OwnerReference
+	if len(export.Spec.Revisions) > 0 {
+		for _, r := range export.Spec.Revisions {
+			tmp := r.DeepCopy()
+			// OwnerRef ensures that Revisions are recognized by controller
+			tmp.OwnerReferences = []metav1.OwnerReference{*kmeta.NewControllerRef(currentConf)}
+			if err = client.CreateRevision(tmp); err != nil {
+				return err
+			}
+		}
+	}
+
+	err = waitIfRequested(client, serviceName, waitFlags, "Importing", "imported", out)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func getConfigurationWithRetry(client clientservingv1.KnServingClient, name string) (*servingv1.Configuration, error) {
+	var conf *servingv1.Configuration
+	var err error
+	err = retry.OnError(retry.DefaultBackoff, func(err error) bool {
```

Comment:
good find ! I think we have several other spots in the client where do a retry, but we do it in a home grown way. I.e to retry on case of a concurrent modification error. Not sure if the k8s utility method could be used there as well.

Reply:
IMO the concurrent modification was the primary purpose to have such util methods, by the following [comment](https://github.com/kubernetes/client-go/blob/1eb2027cd51eefaca0f851d5a2c90f8dd9b2a71d/util/retry/util.go#L68-L75).

Reply:
👍 Let me open a separate issue for moving to the more general way for doing retries.

Reply:
https://github.com/knative/client/issues/1102